pandas/core/groupby/ops.py

```diff
@@ -1260,10 +1260,17 @@ def __iter__(self) -> Iterator:
         starts, ends = lib.generate_slices(self._slabels, self.ngroups)
 
         for start, end in zip(starts, ends):
-            yield self._chop(sdata, slice(start, end))
+            sliced = self._chop(sdata, slice(start, end))
+            if self._sorted_data is self.data:
+                yield sliced.copy()
```

Comment:
can we avoid this?  ideas:

1) with CoW it should be unnecessary right?
2) i think we document that UDFs with side-effects are not supported (or at least caveat emptor)
3) could pre-check  if this is a UDF and not copy in non-UDF cases?
4) doesn't avoid the copy, but defining chop=self._chop, data=self.data, sorted_data=self._sorted_data inside the method should make make this iteration marginally faster in large-ngroups cases

Reply:
For the CoW case, ``deep=None`` inside the copy call would make this a no-op.

https://pandas.pydata.org/docs/user_guide/gotchas.html#mutating-with-user-defined-function-udf-methods says that mutation is not allowed, but the fact that mutation doesn't mess up the DF due to a copy being made is explicitly tested for in tests, right?

e.g. maybe https://github.com/pandas-dev/pandas/blob/23c3dc2c379eb325fd4f33be78338730d3f35731/pandas/tests/groupby/test_apply_mutate.py#L23?

I don't think 4, 5 make a difference (at least looking at a snakeviz profile).

Somehow, the groupby benchmarks are really flaky for me comparing this PR to main, since sometimes I get 10-50% regression, but not when doing a comparison with itself (as a sanity test). The improvements are not flaky, though.

The culprit is probably is duplicate work between ``is_range_indexer`` and ``array_equal_fast`` (in take). Curiously, this doesn't show up in the snakeviz profile...